Document improvements and more flexibility of interface variable check

YaqiWang opened the pull request:
This is a follow-up for PR #21217.

Tag @lindsayad and @GiudGiud. I finished the refactoring of Griffin but debugging that complicated UO, but I feel I can what I had up for you to take a look.

YaqiWang committed:
document improvements and more flexibility of interface variable check for the domain user object #19995

lindsayad commented on framework/include/userobject/DomainUserObject.h:
```diff
@@ -179,6 +191,9 @@ class DomainUserObject : public UserObject,
   const MooseArray<Real> & _coord;
+
+  /// The set of boundaries where variables we wish to evaluate
+  std::map<unsigned int, std::set<BoundaryID>> _interface_var_blocks;
```
Maybe rename this variable

lindsayad commented on framework/src/userobject/DomainUserObject.C:
```diff
@@ -78,35 +76,46 @@ DomainUserObject::DomainUserObject(const InputParameters & parameters)
   mooseAssert(field_var, "We should not be able to return a null variable");
-  _interface_vars.insert(field_var);
+  if (interfaces)
+    _interface_var_blocks[field_var->number()] = *interfaces;
+  else
+    _interface_var_blocks[field_var->number()] = _interface_bnd_ids;
```
I'd rather not store by number. This could be problematic if a user couples in both a nonlinear variable and an aux variable

YaqiWang replied:
Oh, that makes me worry too. Do we have a magic treatment in MOOSE to give them different numbers?

lindsayad replied:
haha no but you can always rely on the pointers being different!

YaqiWang replied:
Decided to just use name string. I think MOOSE made sure the variable names across all systems are unique.

lindsayad reviewed (changes requested):
Overall I like the changes

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1105835/) on 00a0a25 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/21543/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1105863/) on 00a0a25 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>195b98</th>
      <th colspan=4>#21543 00a0a2</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>83.53%</td>
      <td>83.53%</td>
      <td>-0.01%</td>
      <td>84.62%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>77810</td>
      <td>77812</td>
      <td>+2</td>
      <td>22</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15339</td>
      <td>15345</td>
      <td>+6</td>
      <td>4</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/21543/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/21543/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/21543/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/21543/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/21543/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/21543/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/21543/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/21543/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/21543/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/21543/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/21543/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/21543/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/21543/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/21543/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/21543/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/21543/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/21543/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/21543/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/21543/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/21543/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/21543/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/21543/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/21543/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/21543/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/21543/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/21543/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/21543/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 84.62% is less than the suggested 90.0%



This comment will be updated on new commits.

YaqiWang committed:
need to handle the interface that is connected with the domain but the domain is not on its primary side #19995

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1105982/) on 43b7729 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/21543/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/21543/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 195b98388f2ef273e0ebb063bbf6b042be7ed0d3`

YaqiWang commented:
I added one more check in the constructor to make sure all provided interfaces are connected with the domain. I think that is a defensive check and should be good.

YaqiWang commented:
One big change is in `ComputeUserObjectsThread`. The domain user object may not on the primary side of the interface, so I had to change the logic there a bit. I've confirmed that with this PR, I can finish the refactoring in Griffin successfully. My changes in Griffin have been pushed up.

GiudGiud commented on framework/include/materials/ThreeMaterialPropertyInterface.h:
```diff
@@ -13,6 +13,14 @@
+ * This interface is designed currently for DomainUserObject where material properties
+ * on element, face and neighboring face may all required. It can be used in the future
+ * by objects having similar requirements as DomainUserObject. The base interface
+ * TwoMaterialPropertyInterface only provides functions to retrieve material properties
+ * on face and neighboring face.
```
```suggestion
 * TwoMaterialPropertyInterface only provides functions to retrieve material properties
 * on the element and its neighbor
```

GiudGiud commented on framework/include/materials/ThreeMaterialPropertyInterface.h:
```diff
@@ -13,6 +13,14 @@
+
+/**
+ * This interface is designed currently for DomainUserObject where material properties
+ * on element, face and neighboring face may all required. It can be used in the future
```
```suggestion
 * on element, face and neighbor may all required. It can be used in the future
```

GiudGiud reviewed (changes requested):
my current understanding of two material prop interface

GiudGiud commented on framework/doc/content/syntax/UserObjects/index.md:
```diff
@@ -30,8 +30,10 @@ and use the interface without the use of dynamic casts.
    within the mesh on each execution flag.
 - +SideUserObject+: "execute" is called for each +side+, that is on a boundary, within the mesh
```
```suggestion
- +SideUserObject+: "execute" is called for each +side+, that is on a boundary (internal or external), within the mesh
```

GiudGiud commented on framework/include/materials/MaterialPropertyInterface.h:
```diff
@@ -310,6 +310,7 @@ class MaterialPropertyInterface
 
   /**
    * Retrieve the property deduced from the name \p name for the specified \p material_data
+   * that can be defined on element, face and neighboring face.
```
```suggestion
   * that can be defined on element, neighbor and face.
```

YaqiWang replied:
`neighbor` could mean neighboring element, so I did that on purpose to say neighboring face.

GiudGiud replied:
it doesnt do neighboring face? It does neighbor element

GiudGiud replied:
or like neighbor qp probably

YaqiWang replied:
Not exactly, it is the same side but on the neighboring side. Basically for a side, we have `+` and `-` quantities for face and neighboring face respectively.

GiudGiud replied:
it s retrieved on the face qps, but it s only defined on the neighbor elements right? It doesnt have to be, but TwoMatPropInterface is designed for that purposed

YaqiWang replied:
Yes.

lindsayad replied:
I think @YaqiWang's phrasing using the "neighbor face" is the most accurate. I would not want just generic "neighbor" language as that can lead users to think volumetric neighbor

GiudGiud commented on framework/include/materials/ThreeMaterialPropertyInterface.h:
```diff
@@ -13,6 +13,14 @@
+ * TwoMaterialPropertyInterface only provides functions to retrieve material properties
+ * on face and neighboring face.
+ */
 class ThreeMaterialPropertyInterface : public TwoMaterialPropertyInterface
```
also add a docstring to TwoMaterialPropertyInterface please

GiudGiud commented on framework/include/userobject/DomainUserObject.h:
```diff
@@ -28,6 +28,17 @@ class Elem;
+ * We require this parameter for interface evaluations because we want to enforce sanity
+ * checks on coupling variables that are not defined on the domain this user object works on
+ * but only are available on interfaces.
+ * With this user object, evaluations do not have to be put into ElementUserObject,
```
```suggestion
 * With this user object, evaluations that would have to be put into ElementUserObject,
```

GiudGiud commented on framework/include/userobject/DomainUserObject.h:
```diff
@@ -28,6 +28,17 @@ class Elem;
+ * checks on coupling variables that are not defined on the domain this user object works on
+ * but only are available on interfaces.
+ * With this user object, evaluations do not have to be put into ElementUserObject,
+ * SideUserObject, InternalSideUserObject, and InterfaceUserObject separately.
```
```suggestion
 * SideUserObject, InternalSideUserObject, and InterfaceUserObject separately may be combined.
```

GiudGiud commented on framework/include/userobject/DomainUserObject.h:
```diff
@@ -101,10 +112,14 @@ class DomainUserObject : public UserObject,
-   * performing our block-restriction integrity check
+   * performing our block-restriction integrity check.
+   * The argument \p interfaces is optional specifying on what interfaces the variable are expected
+   * to be defined. Default value means that the variable should be defined on all interfaces.
```
let s be more precise. Is this really turning it on for ALL internal sidesets?
or just near the domain

YaqiWang replied:
Replaced with:
```
   * The argument \p interfaces is optional specifying on what interfaces the variable is expected
   * to be available, i.e. the field variable is defined over elements out side of the domain but
   * connecting the subdomain with the interface. Default value means that the variable should be
   * available on all interfaces.
   * Note: a field variable on interface is not required to be defined on the subdomain of this
   *       domain user object.
```

GiudGiud commented on framework/include/userobject/DomainUserObject.h:
```diff
@@ -179,6 +191,9 @@ class DomainUserObject : public UserObject,
   /// RZ, 4piR^2 for spherical
   const MooseArray<Real> & _coord;
+
+  /// The set of boundaries where variables we wish to evaluate
+  std::map<VariableName, std::set<BoundaryID>> _var_interfaces;
```
rework docstring. This isnt a set

GiudGiud commented on framework/include/utils/MooseMeshUtils.h:
```diff
@@ -24,10 +24,13 @@ void changeBoundaryId(MeshBase & mesh,
+                                       const std::vector<BoundaryName> & boundary_name,
+                                       bool generate_unknown);
+
+std::set<BoundaryID> getBoundaryIDSet(const libMesh::MeshBase & mesh,
+                                      const std::vector<BoundaryName> & boundary_name,
+                                      bool generate_unknown);
```
not a fan of these dual set/vector returns.
Do we have them somewhere else?

YaqiWang replied:
Yes, I am using it in Griffin.

GiudGiud commented on framework/src/loops/ComputeUserObjectsThread.C:
```diff
@@ -245,20 +249,22 @@ ComputeUserObjectsThread::onInterface(const Elem * elem, unsigned int side, Boun
-        break;
-      }
+  bool has_domain_objs = false;
+  // we need to check all domain user objects because a domain user object may not active
```
```suggestion
  // we need to check all domain user objects because a domain user object may not be active
```

GiudGiud commented on framework/src/userobject/DomainUserObject.C:
```diff
@@ -64,49 +64,64 @@ DomainUserObject::DomainUserObject(const InputParameters & parameters)
+          // these are blocks connected to our blocks
+          _interface_connected_blocks[interface_bnd_id].insert(interface_connected_block);
       }
+      if (!has_connection)
+        paramError("interface_boundaries",
+                   "Interface with ID ",
+                   interface_bnd_id,
+                   " is not connected with the domain of the domain user object ",
+                   name());
```
ok have my answer for the previous question

GiudGiud commented on framework/src/userobject/DomainUserObject.C:
```diff
@@ -64,49 +64,64 @@ DomainUserObject::DomainUserObject(const InputParameters & parameters)
-                   "' is not defined on the interface connected block '",
-                   _mesh.getSubdomainName(connected_block),
-                   "'");
+    auto & bnd_ids = it->second;
+    for (const auto bnd_id : bnd_ids)
+    {
+      const auto & connected_blocks = _interface_connected_blocks.at(bnd_id);
+      for (const auto & bid : connected_blocks)
+        if (!variable.hasBlocks(bid))
+          mooseError("Variable '",
+                     variable.name(),
+                     "' is not defined on the interface connected block '",
+                     _mesh.getSubdomainName(bid),
+                     "'");
```
but that s part of the point of interfaces right. an interface UO may be the interface between two different quantities, defined on each side

lindsayad replied:
The application developer will specify a parameter like
```c++
  params.addRequiredCoupledVar("v", "The variable on the other side of the interface");
```
and then we retrieve that kind of variable with a `getInterfaceFieldVar` API which marks it as a "special" variable for which we want to check integrity in a different way than the typical `BlockRestrictable` check which is meant for variables defined on the same blocks as the `DomainUserObject`

YaqiWang commented:
The most difficulty part of implementing domain user object is the place related with interface. I reworked the words and checks. @GiudGiud I addressed your comments except `neighboring face` to `neighbor` part.

lindsayad commented on framework/src/userobject/DomainUserObject.C:
```diff
@@ -64,49 +64,74 @@ DomainUserObject::DomainUserObject(const InputParameters & parameters)
+    auto & bnd_ids = it->second;
+    for (const auto bnd_id : bnd_ids)
+    {
+      const auto & connected_blocks = _interface_connected_blocks.at(bnd_id);
```
```suggestion
      const auto & connected_blocks = libmesh_map_find(_interface_connected_blocks, bnd_id);
```

YaqiWang committed:
addressing minor comments #19995

lindsayad reviewed (approved):
would be nice to bump the coverage a little bit, but it's not too bad

YaqiWang commented:
> would be nice to bump the coverage a little bit, but it's not too bad

Possibly just missing run exception tests. I feel this is a feature that can possibly still grow. We can address more specific things if requires come.

YaqiWang commented:
The test is failing in distributed mesh?! Did it fail before?

lindsayad commented:
No it didn't
